### ISSUE TITLE
contributor friendliness metabug

### ISSUE THREAD

**Issue by 6a68** [mozilla/idea-town#5]
What should we do to make idea town maximally contributor friendly with a minimum of process and setup work?

**Comment by meandavejustice** [comment]
I think we are on the right track with the `docker-compose up` command. The `idea-town-app` project is pretty straight forward so far as well, being there is just the gulpfile to run.

**Comment by pdehaan** [comment]
FWIW, I'm a fan of **fxa-content-server**'s [CONTRIBUTING.md](https://github.com/mozilla/fxa-content-server/blob/master/CONTRIBUTING.md) file which outlines PR guidelines and our Git Commit Guidelines (to avoid the wrath of GitCop).

**Comment by 6a68** [comment]
@pdehaan Yeah, CONTRIBUTING.md just came up in a conversation about whether reviewer or committer merges PRs. If reviewer merges, then we can use the same process for volunteers, which is a good rule to follow.

**Comment by 6a68** [comment]
ok, logging won't be setup in the channel until tomorrow, so I"ll excerpt some discussion into this bug for later reference.

### who merges PRs?

Reviewer merges after submitter resubmits.

This ensures PRs are treated the same way, whether or not the submitter has push rights (community contributor friendly).

Discussion:

> 11:03:37 AM < lorchard> Quick github process question: Should I merge my own PRs after a thumbs up? Some projects I’ve been on, submitter should never merge own PRs - reviewer does that. Other projects, it’s expected that submitter merges.
11:04:18 AM < lorchard> Reviewer-merged generally works better for contributors though (since submitters are often not able to merge)
11:06:42 AM < lorchard> On the other hand, submitter-merged works better when you have just one last thing to do (eg. squash & clean up) before it gets away from you :)
11:08:03 AM < lorchard> http://wheningit.tumblr.com/post/32810125420/when-someone-merges-your-pull-request-before-its
11:08:30 AM < JSON_voorhees> I have been wondering about this as well
11:11:01 AM < _6a68> lorchard: good question
11:11:25 AM < _6a68> I tend to roll with either "fix these little things, then merge" or "fix these huge things, then ping me so we can take another look before it lands"
11:11:44 AM < _6a68> reviewer merges seems to be simpler than fine details like that, though
11:11:57 AM < _6a68> do you guys have preferences about the process?
11:12:22 AM < _6a68> How about we start documenting these decisions in a CONTRIBUTING.md file in the main idea-town repo?
11:12:47 AM < lorchard> I’m fine either way… I just don’t want to leave a merge hanging because I figured some one else would, or merge something when I shouldn’t :)
11:12:51 AM < _6a68> I'll get something together and submit a PR and we can have the mother of all bikeshedding discussions in the PR
11:13:03 AM < _6a68> lorchard: lol yeah, let's do reviewer merges - that way contributors use the same model as us
11:14:59 AM < JSON_voorhees> I don't really have an opinion other than there be a consistant thing we do
11:15:09 AM < JSON_voorhees> either way is fine with me though
11:20:21 AM < _6a68> lorchard: speaking of such things, let's simplify the git commit format
11:22:01 AM < lorchard> Yeah, reviewer merges works for me

### git commit message formatting

Decision: Follow [these seven rules](http://chris.beams.io/posts/git-commit/#seven-rules) except:
- Put the github bug-closing "Fix #123: " at the start of the line, to make it easier to correlate commits and issues: `Fix #8675: Derezz the master control program`
- This makes the line longer, so aim for 72 char max, not 50 char max

Discussion:

> 11:49:41 AM < •_6a68> JSON_voorhees | lorchard: sorry, the tooling discussion derailed the git commit log discussin
11:49:49 AM < •_6a68> how about the approach proposed here? http://chris.beams.io/posts/git-commit/
11:50:25 AM < lorchard> Yeah, that’s pretty standard for me
11:50:32 AM < •_6a68> better link: http://chris.beams.io/posts/git-commit/#seven-rules
11:50:33 AM < lorchard> or aspirationally standard :)
11:50:37 AM < •_6a68> haha, me too
11:50:55 AM < •_6a68> so, only thing missing there is where to put "fixes #NN"
11:50:58 AM < lorchard> Though I often try for like 72 chars in subject
11:51:10 AM < •_6a68> I've never been zealous about a short subject, so putting it in there is what I've generally done
11:51:15 AM < lorchard> because 50 is sometimes *way* too short for some one-line explanations 
11:51:19 AM < •_6a68> yeah
11:51:32 AM < lorchard> Though I guess puttng “Closes #123” at the end is fair
11:51:50 AM < lorchard> I’d put “Issue #123: sdfgsdfgsdfg” in the subject, which might also be why 50 chars is too short :)
11:51:53 AM < •_6a68> "Refactor the frob. Fixes #123"? "Closes #123?"
11:52:01 AM < •_6a68> ah nice, I like that
11:52:30 AM < •_6a68> "Issue #345: Specific present-tense commit message"
11:52:38 AM < •_6a68> JSON_voorhees: any thoughts? http://chris.beams.io/posts/git-commit/#seven-rules ?
11:53:44 AM < •_6a68> modified to allow 72 chars and start the line with "Issue #123: "
11:54:25 AM < lorchard> Could also use Fixes #123, or really anything in here https://help.github.com/articles/closing-issues-via-commit-messages/
11:54:44 AM < •_6a68> yeah
11:55:43 AM < •_6a68> "Fix #123: Do The Thing" maybe?
11:58:42 AM < lorchard> Yeah, that’s cool too
11:59:17 AM < JSON_voorhees> yeah I really like that
11:59:48 AM < JSON_voorhees> I don't find the Bug, Feat, Task stuff
11:59:55 AM < JSON_voorhees> i mean I dont MIND
12:00:06 PM < JSON_voorhees> then the fixes #?? in the message body
12:01:27 PM < •_6a68> looks like we're abandoning all that, instead shoving the fixes in the first line, but in imperative tense, to match the tense of the summary:
12:01:32 PM < •_6a68> "Fix #8675: Derezz the master control program"
12:01:46 PM < •_6a68> JSON_voorhees: any objections? If not I'll log the decision in the bug
12:02:16 PM < •_6a68> "Fix" looks a bit weird to me, but reads better
12:02:50 PM < •_6a68> The idea is to keep the bug number easy to find with a skim of the commit history
12:04:34 PM <• _6a68> I went ahead and deleted the gitcop configs for our repos, since it's really built to handle the angular format, and doesn't make it easy to do anything more elaborate without $$
New messages
12:05:21 PM < JSON_voorhees> I'm down with that :)


### git flow: branch naming conventions

Decision: use issue number + '-' + short-dashed-description: `85-uninstall-handling`

Discussion:

> 12:06:35 PM <JSON_voorhees> branch name shouldn't really matter, but a thing that is helpful for myself is to name it 85-uninstall-handling or something similiar
12:06:49 PM <lorchard> Yup, that’s how I usually name my own branches too
12:06:50 PM <JSON_voorhees> Which is super helpful since I can never remember issue numbers
12:06:52 PM <•_6a68> oh yeah, we might as well define a rule there too

### Other stuff (no discussion logs)

- create PR branches on a personal fork, not the main repo (again, same thing a community contributor without push access would do)
- manually create changelogs as needed
  - format TBD, depends on what ops wants to know

**Comment by 6a68** [comment]
I'll put the contributing-related stuff into a CONTRIBUTING.md on the main idea-town repo, and we can point the sub-repos at the main one

**Comment by 6a68** [comment]
Les added an editorconfig dotfile, and it looks pretty cool: lets us avoid having to talk about tabs vs spaces in contributor documentation.

Let's definitely include it in CONTRIBUTING.md
